File: mobile/lib/_services/account_service.dart
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import '../_helpers/api_helper.dart';
import '../_helpers/auth_helper.dart';
import '../_models/dtos.dart';

class AccountService extends ChangeNotifier {
  UserDto? _usuarioAtual;
  bool _carregando = false;

  UserDto? get usuarioAtual => _usuarioAtual;
  bool get carregando => _carregando;
  bool get estaLogado => _usuarioAtual != null;
  bool get isAdmin => _usuarioAtual?.role == 'Admin';

  Future<void> inicializar() async {
    final token = await obterToken();
    final email = await obterEmail();
    final role = await obterRole();
    if (token != null && email != null) {
      _usuarioAtual = UserDto(email: email, token: token, role: role ?? '');
      notifyListeners();
    }
  }

  Future<String?> login(LoginDto model) async {
    _carregando = true;
    notifyListeners();
    try {
      final response = await http.post(
        Uri.parse('${baseUrl}account/login'),
        headers: buildHeaders(null),
        body: json.encode(model.toJson()),
      );
      if (response.statusCode == 200) {
        final user = UserDto.fromJson(json.decode(response.body));
        await salvarToken(user.token, user.email, user.role);
        _usuarioAtual = user;
        return null;
      }
      return 'Usuário ou senha inválidos';
    } catch (e) {
      return 'Erro de conexão com o servidor';
    } finally {
      _carregando = false;
      notifyListeners();
    }
  }

  Future<void> logout() async {
    await limparToken();
    _usuarioAtual = null;
    notifyListeners();
  }

  Future<List<UsuarioDto>?> listarUsuarios() async {
    try {
      final response = await http.get(
        Uri.parse('${baseUrl}account/usuarios'),
        headers: buildHeaders(_usuarioAtual?.token),
      );
      if (response.statusCode == 200) {
        final lista = json.decode(response.body) as List<dynamic>;
        return lista.map((e) => UsuarioDto.fromJson(e)).toList();
      }
      return null;
    } catch (e) {
      return null;
    }
  }

  Future<String?> registrar(RegisterDto model) async {
    try {
      final response = await http.post(
        Uri.parse('${baseUrl}account/registrar'),
        headers: buildHeaders(_usuarioAtual?.token),
        body: json.encode(model.toJson()),
      );
      if (response.statusCode == 200) return null;
      final body = json.decode(response.body);
      return body?.toString() ?? 'Erro ao cadastrar usuário';
    } catch (e) {
      return 'Erro de conexão';
    }
  }

  Future<String?> esqueciSenha(EsqueciSenhaDto model) async {
    try {
      final response = await http.post(
        Uri.parse('${baseUrl}account/esqueci-senha'),
        headers: buildHeaders(null),
        body: json.encode(model.toJson()),
      );
      if (response.statusCode == 200) return null;
      return 'Erro ao processar solicitação';
    } catch (e) {
      return 'Erro de conexão';
    }
  }

  Future<String?> excluirUsuario(String id) async {
    try {
      final response = await http.delete(
        Uri.parse('${baseUrl}account/$id'),
        headers: buildHeaders(_usuarioAtual?.token),
      );
      if (response.statusCode == 204) return null;
      return 'Erro ao desativar usuário';
    } catch (e) {
      return 'Erro de conexão';
    }
  }
}

File: client/src/app/_services/account.service.ts
import { HttpClient } from '@angular/common/http';
import { computed, inject, Injectable, signal } from '@angular/core';
import { map, take } from 'rxjs';
import { LoginDto, RegisterDto, UserDto, UsuarioDto } from '../_models/dtos';
import { environment } from '../../_environment/environment';

@Injectable({ providedIn: 'root' })
export class AccountService {
  private http = inject(HttpClient);
  private baseUrl = environment.apiUrl;

  currentUser = signal<UserDto | null>(this.carregarUsuario());
  isAdmin = computed(() => this.currentUser()?.role === 'Admin');

  private carregarUsuario(): UserDto | null {
    const stored = localStorage.getItem('user');
    return stored ? JSON.parse(stored) : null;
  }

  login(model: LoginDto) {
    return this.http.post<UserDto>(`${this.baseUrl}/account/login`, model).pipe(
      take(1),
      map(user => {
        localStorage.setItem('user', JSON.stringify(user));
        this.currentUser.set(user);
        return user;
      })
    );
  }

  logout() {
    localStorage.removeItem('user');
    this.currentUser.set(null);
  }

  listarUsuarios() {
    return this.http.get<UsuarioDto[]>(`${this.baseUrl}/account/usuarios`).pipe(take(1));
  }

  registrar(model: RegisterDto) {
    return this.http.post<UsuarioDto>(`${this.baseUrl}/account/registrar`, model).pipe(take(1));
  }

  excluirUsuario(id: string) {
    return this.http.delete(`${this.baseUrl}/account/${id}`).pipe(take(1));
  }
}
